LUN-232: Add qr code scanner

SteffiStuffel opened the pull request:
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.


Can be tested with this valid QR-Code:
https://api.qrserver.com/v1/create-qr-code/?size=150x150&data=rXTIfCwz.eOYXwwn486fR5PlKFW9KagLtIzwwBCrE

f1sh1918 commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+          onStatusChange={status => setPermissionDenied(status.cameraStatus === 'NOT_AUTHORIZED')}
+          onCameraReady={() => setPermissionDenied(false)}>
+          {() =>
+            permissionDenied && !requestPermissions ? (
```
i think here is no need for tenary expression and render a fragment just use `&&` and remove  `: <></>`

SteffiStuffel replied:
You are right, though because it is a faac it is needed, but it is not. Removed it.

f1sh1918 commented on src/routes/add-custom-discipline/AddCustomDisciplineScreen.tsx:
```diff
@@ -30,22 +33,31 @@ const Description = styled.Text`
+  letter-spacing: 0.11px;
+  font-family: ${props => props.theme.fonts.contentFontRegular};
+  color: ${prop => prop.theme.colors.lunesBlack};
+  width: 80%;
```
```suggestion
  padding-right: 15px
```

SteffiStuffel replied:
This would not work, as when entering a long code, the icon is moved to the right. Adjusted it nevertheless to width 90% and removed flex.

klinzo commented on android/app/src/main/AndroidManifest.xml:
```diff
@@ -2,6 +2,7 @@
   package="app.lunes">
 
     <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission android:name="android.permission.VIBRATE"/>
```
Don't we need the camera permission as well? https://react-native-camera.github.io/react-native-camera/docs/installation.html#android---other-required-steps

SteffiStuffel replied:
weird, why is this working then?

klinzo replied:
No idea. I think we don't need the vibrate permission though, do we?

SteffiStuffel replied:
No, i removed it.

klinzo replied:
Maybe because it's a debug build but not sure.

klinzo commented on src/routes/add-custom-discipline/AddCustomDisciplineScreen.tsx:
```diff
@@ -30,22 +33,31 @@ const Description = styled.Text`
-  letter-spacing: 0.11px;
-  font-family: ${props => props.theme.fonts.contentFontRegular};
-  color: ${prop => prop.theme.colors.lunesBlack};
+const InputContainer = styled.View<{ errorMessage: string }>`
```
```suggestion
const InputContainer = styled.View<{ error: boolean }>`
```

f1sh1918 commented:
Really nice! Works like a charm. If someone gives another approve i could start with ios part

f1sh1918 reviewed (approved):
Tested on android

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+  background-color: ${props => props.theme.colors.lunesWhite};
+`
+
+const NoAuthDescription = styled.Text`
```
```suggestion
const NotAuthorisedDescription = styled.Text`
```

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+`
+
+interface Props {
+  setVisible: (input: boolean) => void
+  setCode: (input: string) => void
```
```suggestion
  setVisible: (visible: boolean) => void
  setCode: (code: string) => void
```

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+  const appState = useRef(AppState.currentState)
+
+  const [isPressed, setIsPressed] = useState<boolean>(false)
+  const [requestPermissions, setRequestPermissions] = useState<boolean>(false)
```
Maybe:
```suggestion
  const [permissionRequested, setPermissionRequested] = useState<boolean>(false)
```

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+
+  useEffect(() => {
+    const subscription = AppState.addEventListener('change', nextAppState => {
+      if (appState.current.match(/inactive|background/) && nextAppState === 'active') {
```
I know this is just copied from the docs but I think we should not use a regex here as regexes are error prone and remove type safety.
```suggestion
      const inactive = appState.current === 'inactive' || appState.current === 'background'
      if (inactive && nextAppState === 'active') {
```
Not sure if we could also change the check to
```suggestion
      if (appState.current !== 'active' && nextAppState === 'active') {
```

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+      }
+      appState.current = nextAppState
+    })
+    return () => {
+      subscription.remove()
+    }
```
```suggestion
    return subscription.remove
```

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+    }
+  }, [])
+
+  const refresh = () => {
```
```suggestion
  const openSettings = () => {
```

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+          onCameraReady={() => setPermissionDenied(false)}>
+          {() =>
+            permissionDenied && !requestPermissions ? (
+              <NotAuthorisedView>
+                <NoAuthDescription>
+                  {labels.addCustomDiscipline.qrCodeScanner.noAuthorization.description}
+                </NoAuthDescription>
+                <Button
+                  onPress={() => setVisible(false)}
+                  label={labels.addCustomDiscipline.qrCodeScanner.noAuthorization.back}
+                  buttonTheme={BUTTONS_THEME.outlined}
+                />
+                <Button
+                  onPress={() => refresh()}
+                  label={labels.addCustomDiscipline.qrCodeScanner.noAuthorization.grant}
+                  buttonTheme={BUTTONS_THEME.contained}
+                />
+              </NotAuthorisedView>
```
Could we extract this to a separate component?

SteffiStuffel replied:
We can, but it think it is not really needed. This component is only < 150 lines long and only has one responsibility: handling the camera and its permissions. So imo it is totally fine to keep it here.

klinzo replied:
I guess it comes down to personal preference, yes. For me this would not be about the general length of the component (I agree here, < 150 LOC is perfectly fine) but the length of the jsx. I think 40 LOC is really long for jsx and makes it hard to review/work with (or at least for me). As this is just a suggestion, feel free to ignore it. Other opinions welcome as well. I just think we should strive for some modularity.

SteffiStuffel replied:
oke convinced me, moved it.

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+}
+
+const AddCustomDisciplineScreen = ({ setVisible, setCode }: Props): ReactElement => {
+  const appState = useRef(AppState.currentState)
```
Optional: To avoid having to use `.current` all the time you could just `useState` (which is interchangeable here) and `useRef` is only syntactic sugar:
```suggestion
  const [appState, setAppState] = useState(AppState.currentState)
```

SteffiStuffel replied:
Tested, does not work. I guess because setAppState causes a rerender then.

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  useEffect(() => {
+    if (requestPermissions) {
```
Could you explain this? AFAICS `requestPermissions` is only ever set to `true` if the app comes from background to foreground while this is rendered. So if I open the app and navigate here without putting it in the background, the permission is never requested? What am I missing here? :see_no_evil:

SteffiStuffel replied:
Yeah maybe i should add a comment here.
This is only needed when you got to the page, deny camera permissions, click then on settings. When then navigating back from settings this code is executed, leading to a second request for permission. The first one is triggered automatically by RNCamera, therefore no need for us to handle.

Not sure if this is the best solution, but i couldn't work out a better on. There is a `onStatusChange` function of RNCamera, but it does not get called (correctly), i guess because there is nor rerender, when navigating back from settings. I tried a lot of different approaches, but either i got stuck in a infinite rerender cycle or the sencond prompt for permissions was not fired. So I ended up with this solution.

SteffiStuffel replied:
oke maybe this is important to mention: it is only needed when navigate to settings and then select "ask every time", when selecting "allow while using the app" it would work fine without this.

klinzo replied:
Can't we just always request permissions in a useFocusEffect/app state change? Doesn't hurt to request permissions twice, right?

SteffiStuffel replied:
I could not get this working, always leading to a loop, if you can make a working solution with useFocusEffect or anything, I would be happy to see it. For me it always leaded into an endless loop of active / focused and opening and closing the request-dialog.

klinzo replied:
will try and let you know :)

SteffiStuffel replied:
oke more or less just renaming. Thought it would be cool to get rid of the useEffect or listening to the AppState.

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -0,0 +1,133 @@
+          captureAudio={false}
+          onBarCodeRead={onBarCodeRead}
+          testID='camera'
+          onStatusChange={status => setPermissionDenied(status.cameraStatus === 'NOT_AUTHORIZED')}
+          onCameraReady={() => setPermissionDenied(false)}>
```
Do we need both?

klinzo commented on src/constants/labels.json:
```diff
@@ -19,6 +19,13 @@
+    },
+    "qrCodeScanner": {
+      "noAuthorization": {
+        "description": "Feature ist ohne Kamerazugriff nicht verfügbar",
```
Perhaps something like the firefox text makes sense:
![signal-2022-02-15-115521](https://user-images.githubusercontent.com/29095487/154048259-a8ad216b-224e-4f61-91db-564abfa851f9.png)
```suggestion
        "description": "Kamerazugriff zum Scannen von QR-Codes erforderlich. Öffnen Sie die Android-Einstellungen, tippen Sie auf Berechtigungen und tippen Sie auf Erlauben.",
```

klinzo replied:
Text for button should be changed as well then to `Einstellungen` or similar

klinzo reviewed (commented):
Tested on android, works fine!
Minor notes:
* If permissions are not granted and you click on the qr code icon, the screen goes black for a really short time and flickers.
* The white header looks a bit weird imo
* Back button in camera screen does nothing

SteffiStuffel commented on src/routes/add-custom-discipline/AddCustomDisciplineScreen.tsx:
```diff
@@ -30,22 +33,31 @@ const Description = styled.Text`
 `
 
+const StyledTextInput = styled.TextInput`
+  flex-grow: 2;
```
removed completely, not needed.

SteffiStuffel commented:
> Tested on android, works fine! Minor notes:
> 
> * If permissions are not granted and you click on the qr code icon, the screen goes black for a really short time and flickers.

Arg, i though i fixed this. Will have another look at it.

> * Back button in camera screen does nothing

Cannot reproduce? You mean the button above the "Setting / Allow"-Button? Works fine for me.

klinzo commented:
> > 
> > * If permissions are not granted and you click on the qr code icon, the screen goes black for a really short time and flickers.
> 
> Arg, i though i fixed this. Will have another look at it.

You could always request permissions manually on render/focus effect and show nothing until this is completed.

> > * Back button in camera screen does nothing
> 
> Cannot reproduce? You mean the button above the "Setting / Allow"-Button? Works fine for me.

No, I meant the "hardware" back button, not the one in the app.

SteffiStuffel commented on src/routes/add-custom-discipline/__tests__/AddCustomDisciplineScreen.spec.tsx:
```diff
@@ -84,4 +89,14 @@ describe('AddCustomDisciplineScreen', () => {
+    const { findByLabelText } = render(<AddCustomDisciplineScreen navigation={navigation} />, {
+      wrapper: wrapWithTheme
+    })
+    const QRCodeIcon = await findByLabelText('qr-code-scanner')
```
All tests in this file are async, because this component is not rendered until `useReadCustomDisciplines` returns the customDisciplines from AsyncStorage.

SteffiStuffel commented:
> > > * If permissions are not granted and you click on the qr code icon, the screen goes black for a really short time and flickers.
> > 
> > 
> > Arg, i though i fixed this. Will have another look at it.
> 
> You could always request permissions manually on render/focus effect and show nothing until this is completed.
> 
> > > * Back button in camera screen does nothing
> > 
> > 
> > Cannot reproduce? You mean the button above the "Setting / Allow"-Button? Works fine for me.
> 
> No, I meant the "hardware" back button, not the one in the app.

I see, added handling.

klinzo commented on src/routes/add-custom-discipline/components/NotAuthorisedView.tsx:
```diff
@@ -0,0 +1,46 @@
+
+interface Props {
+  setVisible: (visible: boolean) => void
+  openSettings: () => void
```
Optional: Could be moved here.

SteffiStuffel replied:
What could be moved here?

klinzo reviewed (commented):
Hint: Some comments are hidden and not yet addressed.

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -44,32 +44,32 @@ const AddCustomDisciplineScreen = ({ setVisible, setCode }: Props): ReactElement
+      if ((appState.current === 'inactive' || appState.current === 'background') && nextAppState === 'active') {
+        setPermissionRequested(false)
       }
+      console.log('app state changed')
```
```suggestion
```

klinzo commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -44,32 +44,32 @@ const AddCustomDisciplineScreen = ({ setVisible, setCode }: Props): ReactElement
   return (
     <Modal visible transparent animationType='fade' onRequestClose={() => setVisible(false)}>
       <Container>
+        <Text>
+          {permissionRequested} {permissionGranted}
+        </Text>
```
```suggestion
```

SteffiStuffel commented on src/routes/add-custom-discipline/components/QRCodeReaderOverlay.tsx:
```diff
@@ -44,32 +44,32 @@ const AddCustomDisciplineScreen = ({ setVisible, setCode }: Props): ReactElement
   useEffect(() => {
-    if (requestPermissions) {
+    if (!permissionRequested) {
+      console.log('useEffect')
```
arrg, sorry about that

klinzo reviewed (approved):
Nice!
